### ISSUE TITLE
[FR] Application-Specific Passwords (with Working Approach)

### ISSUE THREAD

**Issue by so-rose** [docker-mailserver/docker-mailserver#3198]
# Feature Request
**Tl;dr**: Extra, revocable passwords for users using non-SSO email clients in tandem with SSO approaches.

In my organization, I'd like to keep each user's "real" password (I call this the "master key", not to be confused with dovecot master accounts) far away from them.

I would therefore like to make it easier to create revocable, "application-specific passwords" separate from the "master key" using an opt-in environment variable and `setup`.

## Context
Even once #2713 is implemented, **not all email clients can do OAuth**. Notable examples include K9 for Android.

By making it **easier** (it is totally possible, right now, just not obvious) to create and revoke "application-specific passwords" using `setup`, server-side OAuth adoption will be much easier to encourage.

### Is your Feature Request related to a Problem?
When the user accesses their webmail via our SSO, our backend uses the master key to create a one-time URL that logs the user into Roundcube, without them needing to know the password. When they want to use a traditional email client, however, they currently need this master key, which is problematic (users tend to be terrible at password management).

We are working on implementing the Manual Procedure (described below).

### Describe the Solution you'd like
It's very simple, and is a very slightly modified version of this: https://www.happyassassin.net/posts/2014/08/26/adding-application-specific-passwords-to-dovecot-when-using-system-user-accounts/

**Manual (Current) Procedure**
1. Add the following `passdb` entry **before** the existing one, in `/etc/dovecot/conf.d/auth-passwdfile.inc`:
```
passdb {
  driver = passwd-file
  args = /etc/dovecot/app_specific_passwd
}
```

2. Bind a file `/etc/dovecot/conf.d/app_specific_passwd`:
```
mobile-email@example.com:{SHA512-CRYPT}123456789...::::::user=email@example.com
thunderbird-email@example.com:{SHA512-CRYPT}0123abcd...::::::user=email@example.com
outlook-email@example.com:{SHA512-CRYPT}987654321...::::::user=email@example.com
```

You'll notice that there's also a device-specific username associated with each password, making it easy to identify the behavior of users that are making use of this feature.

3. Implement external logic and/or alerts (ex. cronjob that deletes new entries after a year, anomaly detection on login metrics to alert the admin to misuse) to decide when to make revocations and/or automated expiration.

**Ideal Procedure**
Ideally, managing `app_specific_passwd` would be done through `setup`:
1. Set an environment variable, ex. `DOVECOT_USE_APP_PASS=1`, to enable the extra passdb in `auth-passwdfile.inc` .
2. In the container, run `setup add app-pass <DEVICE_NAME> <EMAIL>`
3. For revocation, run `setup revoke app-pass <DEVICE_NAME> <EMAIL>`

Notes on the `setup` implementation I have in mind:
- It should print a 48-bit base64-encoded password sourced from `/dev/urandom`.
- After generation, it should not be retrievable again.
- `DEVICE_NAME` should pass the regex `^[a-z\-]$`
- If `DOVECOT_USE_APP_PASS` is not set, this `setup` command should fail.

**Bonus: External Logic**
This essentially adds a few security features without making use of them. `docker-mailserver` could implement the most important ones out of the box:
-  **Expiration**: I would love to see this setup script automatically setup app password expiration in `cron`, which allows the mail server admin to force the user to rotate their keys once in awhile (ex. default it to one year).
- **fail2ban Revocation**: Misuse (as defined by fail2ban) of a device-specific password could lead to an automatic revocation of the application-specific password.

### Are you going to implement it?

Yes, because I know the probability of someone else doing it is low and I can learn from it.

*I have hacked my way around this; however, I would like to hear whether this is something others might be interested in, especially considering interest in #2713 and the inability of clients like K9 to use its result.*

### What are you going to contribute??
I would suggest a multi-step process:
1. **Documentation**: Documenting the above approach for now.
2. **setup Implementation**: Making the documented approach easy with `setup`.
3. **Bonuses**: Making use of the security gains from this approach by default.

### Alternatives you've considered
**Waiting for OAuth**
OAuth doesn't quite solve the problem of clients that don't support server-side OAuth.

Moreover, some non-human clients (ex. self-hosted software that sends email, or quick sender scripts) will simply never support OAuth. I'd like to enable OAuth for 95% of cases, and have the backup plan be better than just giving out the one password.

**Giving out the Master Key**
There are several problems with having one shared password across all devices. It's difficult to revoke access for the client that leaked the key. Revoking one leaked key completely disables all devices. Statistical anomaly detection of login patterns is less effective. And so on.

### Who will that Feature be useful to?
See Context.

### What have you done already?
See Manual (Current) Procedure.

**Comment by georglauterbach** [comment]
Sorry for not responding here, but all maintainers were pretty busy due to the release of v12.0.0. Your feature request is interesting, but I fear it may also be out of scope for DMS. If you can estimate the size of the change to the code base of DMS, please let us know, so we can evaluate this further.

**Comment by so-rose** [comment]
Totally fine - and by the way, fantastic work on v12!

I've attempted to thoroughly estimate the impact on the DMS code base of this kind of change (tl;dr it's minimal). Please let me know if I can help in any way!

# Estimate of Changes
I believe the amount of changes would be relatively minimal, and in line with the philosophy of "give the sysadmins easy access to the tools they need".

Here's a summary of what I think would need to be changed in which files (from what I could glean from skimming the code base, which I only minimally understand).

## First: Decisions and Why
**Unique Username for Each App Password**
In the wild, we usually expect to use our normal username with app passwords. The issue is, to do so in dovecot raises a lot of complex decisions about how duplicate usernames should be dealt with. Moreover, having a unique UID makes it infinitely easier to detect abuse related to a particular app password and stop it (fail2ban might even do this already on its own, if it measures by UID).

For users, it's also plenty intuitive (if not more so) that their ex. Thunderbird username is `thunderbird-example@domain.com`. Thus, I'd argue that unique UIDs per application-specific password is the right choice.

**The Extra `passdb`**
This may seem a bit distasteful at first glance, and immediately raise scope questions. After all, there's a great many ways to deal with users and auth in email, and DMS as a project should be rightly concerned about the complexity of this! Even so, I believe it's the only really good way.

App passes in DMS should, in my opinion, follow two rules: "Don't touch the real, existing userdb logic in any way", and "sysadmins should be able to turn app pass support on and off with one button and absolutely no other consequences". Keeping app-pass "users" in a different `passdb` is the absolute easiest way to make sure that the userdb is _never_ touched by app-pass commands (just like aliases never touch the `userdb`), and is also the absolute best way to ensure that all app pass auth can be turned off with one ENV variable (just re-comment the dovecot config for the extra `passdb`).

Keep in mind, this would share the same logic with the standard DMS approach to `passdb` , so there's no new auth logic - only a new file.

## Core Setup
Not a whole lot to do here. Just a new passdb and envvar to enable it.

`/target/dovecot/auth-passwdfile.inc`
```diff
...
+ #passdb {
+ #  driver = passwd-file
+ #  args = /etc/dovecot/app_passwds
+ #}
+
passdb {
  driver = passwd-file
  args = scheme=SHA512-CRYPT username_format=%u /etc/dovecot/userdb
}
...
```

`/target/scripts/startup/setup.d/dovecot.sh`
```diff
...
+ function _setup_dovecot_app_pass
+ {
+     [[ ${DOVECOT_USE_APP_PASS} -eq 1 ]] || return 0
+     sed -i <uncomment_the_new_passdb_lines> /etc/dovecot/conf.d/auth-passwdfile.inc
+    # ...maybe touch /etc/dovecot/app_passwds if it doesn't exist?
+ }
...
```

## User Interface
There isn't much more in terms of LOC; just two new commands for app passwords.

Since app passwords are stored in `passdb` format, there's a strong probability that existing functions like `_manage_accounts` and `_db_entry_remove` can be used directly. Thus, the amount of new logic is potentially very low.

`/target/bin/setup`
```diff
...
    ${LBLUE}COMMAND${RESET} quota ${RED}:=${RESET}
        setup quota ${CYAN}set${RESET} <EMAIL ADDRESS> [<QUOTA>]
        setup quota ${CYAN}del${RESET} <EMAIL ADDRESS>
+ 
+     ${LBLUE}COMMAND${RESET} app-pass ${RED}:=${RESET}
+         setup app-pass ${CYAN}add${RESET} <DEVICE NAME> <EMAIL ADDRESS>
+         setup app-pass ${CYAN}revoke${RESET} <DEVICE NAME> <EMAIL ADDRESS> [EXPIRATION]
+         setup app-pass ${CYAN}revoke-expired${RESET}
///
+ _main
+ # ...corresponding argument delegation to ex. 'addapppass`, `revokeapppass`.
```

`/target/bin/addapppass.sh` / `/target/bin/revokeapppass.sh`
```diff
+ #!/bin/bash
+ 
+ function _main
+ {
+     # ...the usual checks
+     
+     ### Procedure: Append to, or delete from, /etc/dovecot/app_passwds
+ }
+ # ...the usual usage info
```

To clarify the `###` Procedure, I'll take each case on its own.

### Procedure for `/target/bin/addapppass.sh`
Replace the `###` block above with something like the following procedure.
- Grab 48 bytes from `/dev/urandom` (perhaps with `dd`) and run it through `base64`.
- Hash it with `doveadm pw`; this should inherit user's choice of hashing algorithm for other passwords.
- Use `_db_entry_add_or_append` here (assuming that it supports the extra `user=` parameter at the end).

The extremely similar case from `/target/scripts/helpers/database/manage/postfix-accounts.sh` does the hashing part in 4 LOC.
```bash
local PASSWD_HASH
PASSWD_HASH=$(doveadm pw -s SHA512-CRYPT -u "${MAIL_ACCOUNT}" -p "${PASSWD}")
# Early failure above ensures correct operation => Add (create) or Replace (update):
_db_entry_add_or_replace "${DATABASE}" "${MAIL_ACCOUNT}" "${PASSWD_HASH}"
```

### Procedure for `/target/bin/revokeapppass.sh`
All we need is a one-liner `sed` deletion of whichever line starts with `<device>-<email>@example.com`.

### Procedure's Usage of Existing Functions
`_manage_accounts` from `/target/scripts/helpers/database/manage/postfix-accounts.sh` might already be usable for adding app passes, since it already does the hashing work we need.

`_db_entry_remove` might also already be usable for removing app passes, since it already does the kind of `sed` work we need.

In general, since `app_passwds` is a `passdb`, it can potentially be treated it with preexisting DMS helpers. Here's an example of what the file looks like:
```
mobile-email@example.com:{SHA512-CRYPT}123456789...::::::user=email@example.com
thunderbird-email@example.com:{SHA512-CRYPT}0123abcd...::::::user=email@example.com
outlook-email@example.com:{SHA512-CRYPT}987654321...::::::user=email@example.com
```

## Loose Ends
There are a few more details:
- Traditional account deletion should also delete associated app passwords.
- Updating the main user password should potentially revoke all app passwords.
- I'm unsure how it interacts with restrictions.
- App passwords should not be assignable to aliases.
- A way of listing a user's app passwords.

## Documentation and Testing
Needless to say, such a change requires documentation. I'd suggest:
- In Usage, documenting all environment variables and updating the `setup.sh` help text.
- In Usage, putting a page "Application-Specific Passwords" squarely under "Advanced Configuration".

Test cases might include:
- Adding an app pass enables login with the new user/pass, and nothing else.
- Revoking an app pass disables login with the new user/pass, and nothing else.
- Removing and adding a user also deletes all their app passes.

And so on.



# Estimate w/Expiration
The above is already extremely useful, and grants the kind of low-level but user-friendly access that seems to be the scope of DMS.

However, expiration of app passwords is the 'killer security feature' of app-passwords. The whole point of this is to get away from perpetual passwords; the above makes this possible, but requires extra config work from the sysadmin. I'd argue that expiration logic is worth including on the mail server itself, to reduce that barrier to entry.

## Brief Summary of Additional Changes
Implementing expiration on top of the above wouldn't be too much extra work; it would additionally involve:
- To `setup`, adding an optional ` [EXPIRATION]` to `setup app-pass add/revoke` ([ISO8601](https://en.wikipedia.org/wiki/ISO_8601) ex. `2023-04-14T09:09:31+00:00`).
- In the `/target/bin/addapppass.sh` script, using `date` to check that the **optional** expiration date is in the future (there's timezone stuff to be careful of here), and recording the expiration date in a new file `/etc/dovecot/app_passwds_expiry` (under a simple schema like `device-email@domain|<expiry_in_iso8601>`.)
- To `setup`, adding a `setup app-pass revoke-expired` command, which scans `/etc/dovecot/app_passwds_expiry` for expired app-passes and runs `delapppass` on those that have expired.

For auto-expiry, the crontab would now simply run `setup app-pass revoke-expired` ex. every minute. The sysadmin can also, of course, run it whenever they like.
- One could envision an environment variable `DOVECOT_APP_PASS_AUTO_EXPIRE=1m` being used during startup to help build a `crontab`.

Key revocation doesn't need to be precisely scheduled in my opinion; however, that it isn't precise should be documented, so that sysadmins who need a guarantee of time-critical revocation (due to ex. regulations) know to use their own scheduling logic to run `setup app-pass revoke-expired` to fit their needs.

# NOTE on User-Facing Passwords
**NOTE** that it's still only the sysadmin that can create app passwords. As with the rest of DMS, it's still the sysadmin's job to create and update all app passwords as needed (or automate this process as they see fit).

This is to clarify that this is _not to suggest any kind of admin interface or anything like that_. Though, such (seperate) applications may be able to simplify their code due to this work.

**Comment by georglauterbach** [comment]
Thank you very much for the detailed feedback - much appreciated! This actually seems feasible 🚀

I will mark this for the next major release, as it is too packed for the upcoming v12.1.0. We would of course require you to provide the PR, including adjustments to scripts, the documentation & additions to the docs - but fear not, maintainers like me will guide and help you when you have submitted the initial PR.

---

A word of warning: our alias code is the one part of our code base that really needs an overhaul; an overhaul is actually long overdue I think. If you think you can improve the code, maybe to lay a better foundation for your work on app-specific passwords, by all means, go ahead. I will try to assist you as best as I can.

**Comment by so-rose** [comment]
Hi @georglauterbach, apologies for the delay. Thanks for reopening; life-permitting, I'm aiming to get around to a draft PR within a few weeks.
